File: orion-federation-svc/src/repositories/FederationRepository.ts
/**
 * Federation Repository - PostgreSQL Implementation
 *
 * Executor and ExecutorHealth repositories using PostgreSQL.
 * Used by FederationService.
 */

import { DatabasePool } from '../utils/database';

export interface ExecutorEntity {
  id: string;
  cluster_id: string;
  name: string;
  region: string;
  status: string;
  cpu_capacity: number;
  memory_capacity_mb: number;
  cpu_used: number;
  memory_used_mb: number;
  running_jobs: number;
  max_concurrent_jobs: number;
  last_heartbeat: Date | null;
  registered_at: Date;
  labels: Record<string, any>;
}

export interface ExecutorHealthEntity {
  executor_id: string;
  status: string;
  cpu_usage_pct: number;
  memory_usage_pct: number;
  running_jobs: number;
  queue_depth: number;
  last_heartbeat: Date;
  response_time_ms: number;
  errors_last_hour: number;
}

export class ExecutorRepository {
  constructor(private pool: DatabasePool) {}

  async create(data: Partial<ExecutorEntity> & { id: string }): Promise<ExecutorEntity> {
    const result = await this.pool.query(
      `INSERT INTO federation_executors (id, cluster_id, name, region, status, cpu_capacity, memory_capacity_mb, cpu_used, memory_used_mb, running_jobs, max_concurrent_jobs, last_heartbeat, registered_at, labels)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14) RETURNING *`,
      [
        data.id,
        data.cluster_id || '',
        data.name || '',
        data.region || '',
        data.status || 'online',
        data.cpu_capacity || 16,
        data.memory_capacity_mb || 32768,
        data.cpu_used || 0,
        data.memory_used_mb || 0,
        data.running_jobs || 0,
        data.max_concurrent_jobs || 10,
        data.last_heartbeat || new Date(),
        data.registered_at || new Date(),
        JSON.stringify(data.labels || {}),
      ]
    );
    return this.mapRow(result.rows[0]);
  }

  async findById(id: string): Promise<ExecutorEntity | null> {
    const result = await this.pool.query('SELECT * FROM federation_executors WHERE id = $1', [id]);
    return result.rows[0] ? this.mapRow(result.rows[0]) : null;
  }

  async findAll(options?: { limit?: number }): Promise<{ entities: ExecutorEntity[] }> {
    const limit = options?.limit || 100;
    const result = await this.pool.query('SELECT * FROM federation_executors LIMIT $1', [limit]);
    return { entities: result.rows.map(this.mapRow) };
  }

  async findAllActive(): Promise<ExecutorEntity[]> {
    const result = await this.pool.query('SELECT * FROM federation_executors WHERE status = $1', ['online']);
    return result.rows.map(this.mapRow);
  }

  async delete(id: string): Promise<boolean> {
    const result = await this.pool.query('DELETE FROM federation_executors WHERE id = $1', [id]);
    return (result.rowCount ?? 0) > 0;
  }

  async updateHeartbeat(id: string, data: { cpu_used?: number; memory_used_mb?: number; running_jobs?: number }): Promise<boolean> {
    const updates: string[] = ['last_heartbeat = NOW()'];
    const params: unknown[] = [id];
    let paramIndex = 2;

    if (data.cpu_used !== undefined) {
      updates.push(`cpu_used = $${paramIndex++}`);
      params.push(data.cpu_used);
    }
    if (data.memory_used_mb !== undefined) {
      updates.push(`memory_used_mb = $${paramIndex++}`);
      params.push(data.memory_used_mb);
    }
    if (data.running_jobs !== undefined) {
      updates.push(`running_jobs = $${paramIndex++}`);
      params.push(data.running_jobs);
    }

    const result = await this.pool.query(
      `UPDATE federation_executors SET ${updates.join(', ')} WHERE id = $1`,
      params
    );
    return (result.rowCount ?? 0) > 0;
  }

  private mapRow(row: Record<string, unknown>): ExecutorEntity {
    return {
      id: row.id as string,
      cluster_id: row.cluster_id as string,
      name: row.name as string,
      region: row.region as string,
      status: row.status as string,
      cpu_capacity: row.cpu_capacity as number,
      memory_capacity_mb: row.memory_capacity_mb as number,
      cpu_used: row.cpu_used as number,
      memory_used_mb: row.memory_used_mb as number,
      running_jobs: row.running_jobs as number,
      max_concurrent_jobs: row.max_concurrent_jobs as number,
      last_heartbeat: row.last_heartbeat ? new Date(row.last_heartbeat as string) : null,
      registered_at: new Date(row.registered_at as string),
      labels: typeof row.labels === 'string' ? JSON.parse(row.labels) : (row.labels || {}),
    };
  }
}

export class ExecutorHealthRepository {
  constructor(private pool: DatabasePool) {}

  async upsert(data: Partial<ExecutorHealthEntity> & { executor_id: string }): Promise<ExecutorHealthEntity> {
    const result = await this.pool.query(
      `INSERT INTO federation_executor_health (executor_id, status, cpu_usage_pct, memory_usage_pct, running_jobs, queue_depth, last_heartbeat, response_time_ms, errors_last_hour)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
       ON CONFLICT (executor_id) DO UPDATE SET
         status = EXCLUDED.status,
         cpu_usage_pct = EXCLUDED.cpu_usage_pct,
         memory_usage_pct = EXCLUDED.memory_usage_pct,
         running_jobs = EXCLUDED.running_jobs,
         queue_depth = EXCLUDED.queue_depth,
         last_heartbeat = EXCLUDED.last_heartbeat,
         response_time_ms = EXCLUDED.response_time_ms,
         errors_last_hour = EXCLUDED.errors_last_hour
       RETURNING *`,
      [
        data.executor_id,
        data.status || 'healthy',
        data.cpu_usage_pct || 0,
        data.memory_usage_pct || 0,
        data.running_jobs || 0,
        data.queue_depth || 0,
        data.last_heartbeat || new Date(),
        data.response_time_ms || 0,
        data.errors_last_hour || 0,
      ]
    );
    return this.mapRow(result.rows[0]);
  }

  async findByExecutor(executorId: string): Promise<ExecutorHealthEntity | null> {
    const result = await this.pool.query('SELECT * FROM federation_executor_health WHERE executor_id = $1', [executorId]);
    return result.rows[0] ? this.mapRow(result.rows[0]) : null;
  }

  async findAllLatest(): Promise<ExecutorHealthEntity[]> {
    const result = await this.pool.query('SELECT * FROM federation_executor_health');
    return result.rows.map(this.mapRow);
  }

  private mapRow(row: Record<string, unknown>): ExecutorHealthEntity {
    return {
      executor_id: row.executor_id as string,
      status: row.status as string,
      cpu_usage_pct: row.cpu_usage_pct as number,
      memory_usage_pct: row.memory_usage_pct as number,
      running_jobs: row.running_jobs as number,
      queue_depth: row.queue_depth as number,
      last_heartbeat: new Date(row.last_heartbeat as string),
      response_time_ms: row.response_time_ms as number,
      errors_last_hour: row.errors_last_hour as number,
    };
  }
}

File: orion-platform-svc-go/internal/federation/repository/federation_repository.go
package repository

import (
	"context"
	"database/sql"
	"errors"
	"orion/platform-svc-go/internal/federation/models"
	"github.com/jmoiron/sqlx"
)

type Repository struct {
	db *sqlx.DB
}

func NewRepository(db *sqlx.DB) *Repository {
	return &Repository{db: db}
}

// ---------------------------------------------------------------------------
// Federation Config
// ---------------------------------------------------------------------------

func (r *Repository) CreateFederationConfig(ctx context.Context, c *models.FederationConfig) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO federation_configs (id, tenant_id, name, description, clusters, strategy, status, metadata, created_at, updated_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,NOW(),NOW())`,
		c.ID, c.TenantID, c.Name, c.Description, c.Clusters, c.Strategy, c.Status, c.Metadata,
	)
	return err
}

func (r *Repository) GetFederationConfig(ctx context.Context, tenantID, id string) (*models.FederationConfig, error) {
	var c models.FederationConfig
	err := r.db.GetContext(ctx, &c, `SELECT * FROM federation_configs WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, err
	}
	return &c, nil
}

func (r *Repository) ListFederationConfigs(ctx context.Context, tenantID string) ([]models.FederationConfig, error) {
	var items []models.FederationConfig
	err := r.db.SelectContext(ctx, &items, `SELECT * FROM federation_configs WHERE tenant_id=$1 ORDER BY created_at DESC`, tenantID)
	return items, err
}

func (r *Repository) UpdateFederationConfig(ctx context.Context, c *models.FederationConfig) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE federation_configs SET name=$3, description=$4, clusters=$5, strategy=$6, status=$7, metadata=$8, updated_at=NOW() WHERE id=$1 AND tenant_id=$2`,
		c.ID, c.TenantID, c.Name, c.Description, c.Clusters, c.Strategy, c.Status, c.Metadata,
	)
	return err
}

func (r *Repository) DeleteFederationConfig(ctx context.Context, tenantID, id string) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM federation_configs WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	return err
}

// ---------------------------------------------------------------------------
// Executor
// ---------------------------------------------------------------------------

func (r *Repository) CreateExecutor(ctx context.Context, e *models.Executor) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO executors (id, tenant_id, cluster_id, name, region, status, cpu_capacity, memory_capacity_mb, cpu_used, memory_used_mb, running_jobs, max_concurrent_jobs, last_heartbeat, registered_at, labels) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,NOW(),$14)`,
		e.ID, e.TenantID, e.ClusterID, e.Name, e.Region, e.Status,
		e.CPUCapacity, e.MemoryCapacityMB, e.CPUUsed, e.MemoryUsedMB,
		e.RunningJobs, e.MaxConcurrentJobs, e.LastHeartbeat, e.Labels,
	)
	return err
}

func (r *Repository) GetExecutor(ctx context.Context, tenantID, id string) (*models.Executor, error) {
	var e models.Executor
	err := r.db.GetContext(ctx, &e, `SELECT * FROM executors WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, err
	}
	return &e, nil
}

func (r *Repository) GetExecutorWithHealth(ctx context.Context, tenantID, id string) (*models.Executor, *models.ExecutorHealth, error) {
	var e models.Executor
	err := r.db.GetContext(ctx, &e, `SELECT * FROM executors WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, nil, err
	}
	var h models.ExecutorHealth
	err = r.db.GetContext(ctx, &h, `SELECT * FROM executor_health WHERE executor_id=$1 ORDER BY last_heartbeat DESC LIMIT 1`, id)
	if err != nil && !errors.Is(err, sql.ErrNoRows) {
		return &e, nil, err
	}
	return &e, &h, nil
}

var nullExec = (*models.Executor)(nil)

func (r *Repository) ListExecutors(ctx context.Context, tenantID string, offset, limit int) ([]models.Executor, error) {
	var items []models.Executor
	err := r.db.SelectContext(ctx, &items,
		`SELECT * FROM executors WHERE tenant_id=$1 ORDER BY registered_at DESC OFFSET $2 LIMIT $3`,
		tenantID, offset, limit)
	return items, err
}

func (r *Repository) ListActiveExecutors(ctx context.Context, tenantID string) ([]models.Executor, error) {
	var items []models.Executor
	err := r.db.SelectContext(ctx, &items,
		`SELECT * FROM executors WHERE tenant_id=$1 AND status='online' ORDER BY registered_at DESC`,
		tenantID)
	return items, err
}

func (r *Repository) UpdateExecutor(ctx context.Context, e *models.Executor) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE executors SET name=$3, region=$4, status=$5, cpu_capacity=$6, memory_capacity_mb=$7, cpu_used=$8, memory_used_mb=$9, running_jobs=$10, last_heartbeat=NOW() WHERE id=$1 AND tenant_id=$2`,
		e.ID, e.TenantID, e.Name, e.Region, e.Status,
		e.CPUCapacity, e.MemoryCapacityMB, e.CPUUsed, e.MemoryUsedMB, e.RunningJobs,
	)
	return err
}

func (r *Repository) UpdateExecutorHeartbeat(ctx context.Context, id, tenantID string, cpuUsed, memUsedMB float64, runningJobs, responseTimeMs int) (*models.Executor, error) {
	var e models.Executor
	err := r.db.GetContext(ctx, &e, `SELECT * FROM executors WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, err
	}
	e.CPUUsed = cpuUsed
	e.MemoryUsedMB = memUsedMB
	e.RunningJobs = runningJobs
	_, err = r.db.ExecContext(ctx,
		`UPDATE executors SET cpu_used=$3, memory_used_mb=$4, running_jobs=$5, last_heartbeat=NOW() WHERE id=$1 AND tenant_id=$2`,
		id, tenantID, cpuUsed, memUsedMB, runningJobs,
	)
	return &e, err
}

func (r *Repository) DeleteExecutor(ctx context.Context, tenantID, id string) (bool, error) {
	res, err := r.db.ExecContext(ctx, `DELETE FROM executors WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return false, err
	}
	rows, _ := res.RowsAffected()
	return rows > 0, nil
}

// ExecutorHealth
func (r *Repository) UpsertExecutorHealth(ctx context.Context, h *models.ExecutorHealth) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO executor_health (executor_id, status, cpu_usage_pct, memory_usage_pct, running_jobs, queue_depth, last_heartbeat, response_time_ms, errors_last_hour) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9) ON CONFLICT (executor_id) DO UPDATE SET status=$2, cpu_usage_pct=$3, memory_usage_pct=$4, running_jobs=$5, queue_depth=$6, last_heartbeat=$7, response_time_ms=$8, errors_last_hour=$9`,
		h.ExecutorID, h.Status, h.CPUUsagePct, h.MemoryUsagePct, h.RunningJobs, h.QueueDepth, h.LastHeartbeat, h.ResponseTimeMs, h.ErrorsLastHour,
	)
	return err
}

func (r *Repository) ListExecutorHealth(ctx context.Context, tenantID string) ([]models.ExecutorHealth, error) {
	var items []models.ExecutorHealth
	err := r.db.SelectContext(ctx, &items,
		`SELECT eh.* FROM executor_health eh JOIN executors ex ON eh.executor_id=ex.id WHERE ex.tenant_id=$1`,
		tenantID)
	return items, err
}

// ---------------------------------------------------------------------------
// Scheduling Policy
// ---------------------------------------------------------------------------

func (r *Repository) CreateSchedulingPolicy(ctx context.Context, p *models.SchedulingPolicy) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO scheduling_policies (id, tenant_id, name, description, strategy, rules, status, created_at, updated_at) VALUES ($1,$2,$3,$4,$5,$6,$7,NOW(),NOW())`,
		p.ID, p.TenantID, p.Name, p.Description, p.Strategy, p.Rules, p.Status,
	)
	return err
}

func (r *Repository) ListSchedulingPolicies(ctx context.Context, tenantID string) ([]models.SchedulingPolicy, error) {
	var items []models.SchedulingPolicy
	err := r.db.SelectContext(ctx, &items,
		`SELECT * FROM scheduling_policies WHERE tenant_id=$1 ORDER BY created_at DESC`,
		tenantID)
	return items, err
}

// ---------------------------------------------------------------------------
// Cross-Cluster Job
// ---------------------------------------------------------------------------

func (r *Repository) CreateCrossClusterJob(ctx context.Context, j *models.CrossClusterJob) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO cross_cluster_jobs (id, tenant_id, name, spec, target_clusters, status, scheduled_at, completed_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8)`,
		j.ID, j.TenantID, j.Name, j.Spec, j.TargetClusters, j.Status, j.ScheduledAt, j.CompletedAt,
	)
	return err
}

// ---------------------------------------------------------------------------
// Resource Pool
// ---------------------------------------------------------------------------

func (r *Repository) CreateResourcePool(ctx context.Context, p *models.ResourcePool) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO resource_pools (id, tenant_id, name, description, cluster_id, cpu, memory, used_cpu, used_memory, status, created_at) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,NOW())`,
		p.ID, p.TenantID, p.Name, p.Description, p.ClusterID, p.CPU, p.Memory, p.UsedCPU, p.UsedMemory, p.Status,
	)
	return err
}

func (r *Repository) GetResourcePool(ctx context.Context, tenantID, id string) (*models.ResourcePool, error) {
	var p models.ResourcePool
	err := r.db.GetContext(ctx, &p, `SELECT * FROM resource_pools WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, err
	}
	return &p, nil
}

func (r *Repository) Create(ctx context.Context, d *models.FederatedCluster) error {
	_, err := r.db.ExecContext(ctx,
		`INSERT INTO federated_clusters (id, tenant_id, name, peer_url, protocol, status, config, created_at) VALUES ($1,$2,$3,$4,$5,$6,$7,NOW())`,
		d.ID, d.TenantID, d.Name, d.PeerURL, d.Protocol, d.Status, d.Config,
	)
	return err
}

func (r *Repository) List(ctx context.Context, tenantID string, offset, limit int) ([]models.FederatedCluster, error) {
	var items []models.FederatedCluster
	err := r.db.SelectContext(ctx, &items, `SELECT * FROM federated_clusters WHERE tenant_id=$1 ORDER BY created_at DESC OFFSET $2 LIMIT $3`, tenantID, offset, limit)
	return items, err
}

func (r *Repository) GetByID(ctx context.Context, tenantID, id string) (*models.FederatedCluster, error) {
	var d models.FederatedCluster
	err := r.db.GetContext(ctx, &d, `SELECT * FROM federated_clusters WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil { return nil, err }
	return &d, nil
}

func (r *Repository) Delete(ctx context.Context, tenantID, id string) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM federated_clusters WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	return err
}

func (r *Repository) Count(ctx context.Context, tenantID string) (int, error) {
	var count int
	err := r.db.GetContext(ctx, &count, `SELECT COUNT(*) FROM federated_clusters WHERE tenant_id=$1`, tenantID)
	return count, err
}

func (r *Repository) Update(ctx context.Context, d *models.FederatedCluster) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE federated_clusters SET name=$3, peer_url=$4, protocol=$5, status=$6, config=$7, last_sync=NOW() WHERE id=$1 AND tenant_id=$2`,
		d.ID, d.TenantID, d.Name, d.PeerURL, d.Protocol, d.Status, d.Config,
	)
	return err
}
